### DISCUSSION TITLE
sympy.physics.mechanics. Kane's fr, frstar

### DISCUSSION THREAD

**Discussion by Peter230655** [sympy/sympy#26252] [Q&A]
Below is a simple program to generate the EOMs of a single 3D pendulum, modelled as a rigid body, swinging around O, the fixed point in the inertial frame N.
I came across this: _frstar_ contains the **accelerations** of the generalized coordinates. As per my understanding, this should not be so.
Also 
1. If I form _force = KM.forcing_full,_ force does **not** contain any accelerations.

2. This does not seem to happen with 2D simulations.
    Explanation: I have played around with many 2D and 3D simulations, but never pad attention to fr, frstar, as I needed the mass matrix and the force vector to integrate numerically, using _solve_ivp_. Only when I started to play with _opty_ recently. did I use fr, frstar. I started with 2D simulations with _opty_, and there was no problem. Only when a 3D problem did not work, a hint from @tjstienstra gave me the idea of looking at fr, frstar, and I noticed these acceleratrions in fstar.

Any help would the greatly appreciated!

I think, forming rhs = $massmatrix^{-1} \cdot forcevektor$ and then putting EOM = rhs - state_vector should give the EOM as required by opty. Is this correct?
(But calculating rhs symbolically is not feasible for larger simulations as per my experinece (?) )

 

```
import sympy.physics.mechanics as me
import sympy as sm

N = me.ReferenceFrame('N')                                                                                           
O = me.Point('O')                                                                                                         
O.set_vel(N, 0)                                                                                                        
t = me.dynamicsymbols._t

A0 = me.ReferenceFrame('A0')                                                                                  
P0 = me.Point('P0')                                                             
q0x, q0y, q0z = me.dynamicsymbols('q0x q0y q0z')    
u0x, u0y, u0z = me.dynamicsymbols('u0x, u0y, u0z')                                                                                                                                                    
Tx, Ty = sm.symbols('Tx Ty')                                                                                       

l, m, g, reibung, iXX, iYY, iZZ = sm.symbols('l, m, g, reibung, iXX, iYY, iZZ')                                                                              

A0.orient_body_fixed(N, (q0z, q0x, q0y), '312')
rot = A0.ang_vel_in(N)
A0.set_ang_vel(N, u0x*A0.x + u0y*A0.y + u0z*A0.z)
rot1 = A0.ang_vel_in(N)       

P0.set_pos(O, l*A0.z)
P0.v2pt_theory(O, N, A0)  

I = me.inertia(A0, iXX, iYY, iZZ)       
BODY = [me.RigidBody('P0a', P0, A0, m,(I, P0))]  

FL = [(P0, -m*g*N.z), (A0, Tx*A0.x + Ty*A0.y)]                                     
kd = sm.Matrix([me.dot(rot - rot1, uv) for uv in N])                 

q_ind = [q0x, q0y, q0z]
u_ind = [u0x, u0y, u0z]
 
KM = me.KanesMethod(N, q_ind=q_ind, u_ind=u_ind, kd_eqs=kd)
(fr, frstar) = KM.kanes_equations(BODY, FL)
EOM = kd.col_join(fr + frstar) 
print('OEM DS', me.find_dynamicsymbols(EOM), '\n')

force = KM.forcing_full
MM = KM.mass_matrix_full

print(MM)
print('MM:', me.find_dynamicsymbols(MM))
print('force:', me.find_dynamicsymbols(force))
print('fr:', me.find_dynamicsymbols(fr), '\n')
print('frstar:', me.find_dynamicsymbols(frstar), '\n' )
print('kd:', me.find_dynamicsymbols(kd))    
```

**Comment by moorepants** [comment]
By definition, $F_r^*$ will contain accelerations (double time derivatives of generalized coordinates).

**Reply by moorepants** [reply]
See: https://moorepants.github.io/learn-multibody-dynamics/eom.html

**Comment by Peter230655** [comment]
Fair enough! :-(
I did not think right!
I had hoped, this was related to my error with opty:

ValueError                                Traceback (most recent call last)
Cell In[14], [line 41](vscode-notebook-cell:?execution_count=14&line=41)
     [39](vscode-notebook-cell:?execution_count=14&line=39)     prob.add_option('max_iter', 3000)           # defaul is 3000
     [40](vscode-notebook-cell:?execution_count=14&line=40) # Find the optimal solution.
---> [41](vscode-notebook-cell:?execution_count=14&line=41)     solution, info = prob.solve(initial_guess)
     [42](vscode-notebook-cell:?execution_count=14&line=42)     print('message from optimizer:', info['status_msg'])
     [43](vscode-notebook-cell:?execution_count=14&line=43)     initial_guess = solution

File [c:\Users\Peter\anaconda3\envs\sympy-dev\Lib\site-packages\cyipopt\cython\ipopt_wrapper.pyx:621](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/cyipopt/cython/ipopt_wrapper.pyx:621), in ipopt_wrapper.Problem.solve()

ValueError: Wrong length of x0.

But I have to look somewhere else. I just have no idea, where to look at the moment.

**Comment by Peter230655** [comment]
I found the problem. I had declared the specified_symbols as _sm.symbols_ instead of as _me.dynamicsymbols_ when I set up Kane's equations.
